src/JustEat.StatsD/Buffered/BufferBasedStatsDPublisher.cs

```diff
@@ -0,0 +1,184 @@
+        {
+            Increment(value > 0 ? -value : value, sampleRate, bucket);
+        }
+
+        public void Decrement(long value, double sampleRate, params string[] buckets)
+        {
+            if (buckets == null)
+            {
+                return;
+            }
+
+            foreach (var bucket in buckets)
+            {
+                Increment(value > 0 ? -value : value, sampleRate, bucket);
+            }
+        }
+
+        public void Gauge(double value, string bucket)
+        {
+            SendMessage(DefaultSampleRate, StatsDMessage.Gauge(value, bucket));
+        }
+
+        public void Gauge(double value, string bucket, DateTime timestamp)
+        {
+            Gauge(value, bucket);
+        }
+
+        public void Gauge(long value, string bucket)
+        {
+            Gauge((double) value, bucket);
+        }
+
+        public void Gauge(long value, string bucket, DateTime timestamp)
+        {
+            Gauge(value, bucket);
+        }
+
+        public void Timing(TimeSpan duration, string bucket)
+        {
+            Timing(duration.Ticks, bucket);
+        }
+
+        public void Timing(TimeSpan duration, double sampleRate, string bucket)
+        {
+            Timing(duration.Ticks, sampleRate, bucket);
+        }
+
+        public void Timing(long duration, string bucket)
+        {
+            Timing(duration, DefaultSampleRate, bucket);
+        }
+
+        public void Timing(long duration, double sampleRate, string bucket)
+        {
+            var msg = StatsDMessage.Timing(duration, bucket);
+            SendMessage(sampleRate, msg);
+        }
+
+        public void MarkEvent(string name)
+        {
+            Increment(name);
+        }
+
+        private void SendMessage(double sampleRate, in StatsDMessage msg)
+        {
+            bool shouldSendMessage = sampleRate >= DefaultSampleRate || sampleRate > Random.NextDouble();
+
+            if (!shouldSendMessage)
+            {
+                return;
+            }
+
+            try
+            {
+                var buffer = Buffer;
+
+                if (_formatter.TryFormat(msg, sampleRate, buffer, out int written))
+                {
+                    _transport.Send(new ArraySegment<byte>(buffer, 0, written));
+                }
+                else
+                {
+                    var newSize = _formatter.GetMaxBufferSize(msg);
+                    
+                    _buffer = new byte[newSize];
+
+                    if (_formatter.TryFormat(msg, sampleRate, _buffer, out written))
+                    {
+                        _transport.Send(new ArraySegment<byte>(_buffer, 0, written));
+                    }
+                    else
+                    {
+                        // so we was not able to write to resized buffer
+                        // that means there is a bug in formatter
+                        throw new Exception("Utf8 Formatting Error. This is a bug." +
```

Comment:
I'm not sure about throwing vague exceptions to users going "oh, there's a bug".

![image](https://user-images.githubusercontent.com/1439341/45454491-0489cc00-b6dc-11e8-9c44-fc45325ce746.png)